File: Data Structures & Algorithms/counting-bits/submission-0.java
class Solution {
    public int[] countBits(int n) {
        int[] res = new int[n+1];
        int[] memo = new int[n+1];
        
        for (int i = 1; i <= n; i++) {
            res[i] = countBitsOnce(i, memo);
        }
        return res;
    }

    private int countBitsOnce(int n, int[] memo) {
        if (n == 0) return 0;
        if (memo[n] != 0) return memo[n];

        int res = n % 2 + countBitsOnce(n / 2, memo);
        memo[n] = res;
        return res;
    }

}

File: Data Structures & Algorithms/string-encode-and-decode/submission-0.py
class Solution:

    def encode(self, strs: List[str]) -> str:
        res = ""
        for s in strs:
            res += str(len(s)) + '@'
            res += s
        return res
    
    def decode(self, s: str) -> List[str]:
        res = []
        i = 0
        while i < len(s):
            n = ""
            while s[i].isdigit():
                n += s[i]
                i += 1
            n = int(n)
            # s[i] = '@'
            i += 1
            thisStr = ""
            for _ in range(n):
                thisStr += s[i]
                i += 1
            res.append(thisStr)
        return res


